renderers/opengl/msdf.lisp

```diff
@@ -667,19 +667,44 @@
           (when pos
             (alloy:move-to pos component)))))))
 
-(defclass selection (opengl::polygon simple:selection) ())
+(defclass selection (opengl::polygon simple:selection)
+  ((%line-rectangles :reader line-rectangles
+                     :initform (make-array 0 :adjustable t :fill-pointer 0))))
 
 (defmethod simple:selection ((renderer renderer) (text text) (start integer) (end integer) &rest initargs)
   (apply #'make-instance 'selection :text text :start start :end end initargs))
 
 (defmethod shared-initialize ((selection selection) slots &key)
   (call-next-method)
-  (if (/= (simple:start selection) (simple:end selection))
-      (let ((start (compute-cursor-location (simple:text-object selection) (simple:start selection)))
-            (end (compute-cursor-location (simple:text-object selection) (simple:end selection))))
-        (setf (simple:points selection)
-              (list (alloy:px-point (alloy:pxx start) (alloy:pxy start))
-                    (alloy:px-point (alloy:pxx end) (alloy:pxy end))
-                    (alloy:px-point (alloy:pxx end) (+ (alloy:pxy end) (alloy:pxh end)))
-                    (alloy:px-point (alloy:pxx start) (+ (alloy:pxy end) (alloy:pxh end))))))
-      (setf (simple:points selection) ())))
+  (let ((line-rectangles (line-rectangles selection))
+        (start (simple:start selection))
+        (end (simple:end selection)))
+    (setf (fill-pointer line-rectangles) 0)
+    (when (/= start end)
+      (let* ((text (simple:text-object selection))
+             (string (alloy:text text)))
+        (flet ((add-text-line (start-index end-index)
+                 (let* ((start-rect (compute-cursor-location text start-index))
+                        (end-rect (compute-cursor-location text end-index))
+                        (line-height (alloy:pxh start-rect))
+                        (x1 (alloy:pxx start-rect))
+                        (x2 (alloy:pxx end-rect))
+                        (y1 (alloy:pxy start-rect))
+                        (y2 (+ y1 line-height))
+                        (points (list (alloy:px-point x2 y1)
+                                      (alloy:px-point x2 y2)
+                                      (alloy:px-point x1 y2)
+                                      (alloy:px-point x1 y1)))
+                        (rect (make-instance 'opengl::polygon :name 'line
+                                                              :points points)))
```

Comment:
a polygon is going to be a lot less efficient to render than a `filled-rectangle`, so using that would be better, though even better would be to not use delegate shapes at all, and instead only store the bottomleft corner point and width/height size, then in the `render-direct` method do mostly what the `filled-rectangle` does, since then we can eliminate the repeated setup calls and reduce it down to updating two uniforms (`size` and `transform`) and a draw call per rectangle.